compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1619,34 +1619,56 @@ pub type PlaceholderConst = Placeholder<BoundVar>;
 /// know their corresponding parameters. We (ab)use this by
 /// calling `type_of(param_did)` for these arguments.
 ///
+/// # Examples
+///
 /// ```rust
 /// #![feature(const_generics)]
 ///
 /// struct A;
 /// impl A {
-///     fn foo<const N: usize>(&self) -> usize { N }
+///     fn foo<const N: usize>(&self) -> [u8; N] { [0; N] }
+///     //           ^ const parameter
 /// }
 /// struct B;
 /// impl B {
-///     fn foo<const N: u8>(&self) -> usize { 42 }
+///     fn foo<const M: u8>(&self) -> usize { 42 }
 /// }
 ///
 /// fn main() {
 ///     let a = A;
-///     a.foo::<7>();
+///     let _b = a.foo::<7>();
+///     //               ^ const argument
 /// }
 /// ```
+///
+/// The important section here is the call `a.foo::<7>()`. We do not know which
+/// `foo` is used until we know the type of `a`.
+///
+/// We only know the type of `a` once we are inside of `typeck(main_def_id)`.
+/// During that we also want to know the type of `_b` which depends on the
+/// the value of the const argument.
+///
+/// To evaluate that const argument we need to know its type, requiring us to
+/// resolve `foo` as it can otherwise be either `usize` of `u8` in this example.
+/// Resolving `foo` once again requires `typeck(main_def_id)` though,
+/// which results in a cycle.
+///
+/// We only care about the const argument once we know which paramter they instantiate.
+/// So instead of using `type_of(const_arg)` we can use `type_of(const_param)` which does
+/// not depend on `typeck`. This however requires us to pass the `const_param`
+/// to every place which potentionally needs the type of a const argument and gets used
+/// during `typeck`.
 #[derive(Copy, Clone, Debug, TypeFoldable, Lift, TyEncodable, TyDecodable)]
 #[derive(PartialEq, Eq, PartialOrd, Ord)]
 #[derive(Hash, HashStable)]
 pub struct WithOptConstParam<T> {
     pub did: T,
-    /// The `DefId` of the corresponding generic paramter in case `did` is
+    /// The `DefId` of the corresponding generic parameter in case `did` is
     /// a const argument.
     ///
     /// Note that even if `did` is a const argument, this may still be `None`.
     /// All queries taking `WithOptConstParam` start by calling `tcx.opt_const_param_of(def.did)`
-    /// to potentially update `param_did` in case it `None`.
+    /// to potentially update `param_did` in case it is `None`.
```

Comment:
```suggestion
    /// to potentially update `param_did` in the case it is `None`.
```
This made it sound like `opt_const_param_of` is always called.